Update the theme php architecture

perruche opened the pull request:
## Added
Add an autoloaded `app` folder to hold all the php files (excluding WordPress template files such as single.php)

```
|- app
|-- Managers
|-- Models
|-- Repositories
|-- Services
```

#### Managers
Use Managers to extend theme, plugins, Wordpress to your needs.
- `ThemeManager` for Theme related functions
- `AssetsManager` for enquing styles & scripts
- `TwigManager` for managing Twig extensions and functions
- `WordPressManager` to manage global WordPress features
- `ACFManager` to manage ACF fields, groups an option pages
- `CustomPostTypesManager` to manage CPT
- `TaxonomiesManager` to manage taxonomies

#### Models
Use Models to extend the `TimberPost` object to fit post type needs.

#### Repositories
Use Repositories to put query related logic.

#### Services
Use Services to put common functionality that you want to be used in various places. 

---


## Todo
###### Managers

- [x] Add ACFManager
- [x] Add CustomPostTypesManager
- [x] Add TaxonomiesManager

###### Models
- [x] Add Model class exemple

###### Repositories
- [x] Add default Repository class 

###### Services
- [ ] Add a Mailcatching Service 

###### Documentation
- [x] Add documentation in a README.md file for `Managers` folder
- [x] Add documentation in a README.md file for `Models` folder
- [x] Add documentation in a README.md file for `Repositories`folder
- [ ] Add documentation in a README.md file for `Services`folder

perruche committed:
Add folders for the new PHP architecture

- Managers, hold theme, WordPress and plugins setup
- Models, hold and extend data
- Repositories, hold and extend query related logic
- Services, hold low lying abstract features

perruche committed:
Add the app folder to composer autoloader

perruche committed:
Update functions.php to use the Managers

Add some basic managers:
- ThemeManager, for theme related stuff who is initializing all the
other ones
- AssetsManager, to bootstrap Studiometa\WP\Assets feature
- WordPressManager, for WordPress core related stuff

perruche committed:
Add a TwigManager

perruche committed:
Add CustomPostTypesManager and TaxonomiesManager

perruche committed:
Update phpcs rules to ignore some filenaming rules in the app folder

titouanmathis reviewed (commented):
Nice addition 👍

perruche committed:
Add ACFManager

perruche committed:
Add a dummy Model example

perruche committed:
Add Repository

titouanmathis commented on template/web/wp-content/themes/studiometa/app/Managers/ACFManager.php:
```diff
@@ -0,0 +1,84 @@
+	 *
+	 * @return void
+	 */
+	public function register_acf_exemple_group() {
```
Should be `register_acf_example_group` 😉

titouanmathis commented on template/web/wp-content/themes/studiometa/app/Managers/ACFManager.php:
```diff
@@ -0,0 +1,84 @@
+	public function register_acf_exemple_group() {
+		acf_add_local_field_group(
+			array(
+				'key'      => 'group_studiometa_exemple_group',
```
`group_studiometa_example_group`

titouanmathis commented on template/web/wp-content/themes/studiometa/app/Repositories/PostTypeRepository.php:
```diff
@@ -0,0 +1,93 @@
+	public function posts_by_category_slug( $slug, $limit = 10, $exclude = array(), $paged = 0 ) {
+
+		// Set sane defaults so we don't do full table scans.
+		if ( $limit <= 0 || $limit > 100 ) {
```
Maybe throw an error or warning if the `$limit` parameter is bigger than 100 to let the developer know it should not be done?

```php
if ( $limit <= 0 || $limit > 100 ) {
  trigger_error( 'Some message to warn...', E_USER_WARNING );
  $limit = 100;
}

titouanmathis commented on template/web/wp-content/themes/studiometa/app/Repositories/PostTypeRepository.php:
```diff
@@ -0,0 +1,93 @@
+			$limit = 100;
+		}
+
+		// Note the + symbol. See https://codex.wordpress.org/Class_Reference/WP_Query#Category_Parameters.
```
The URL is not valid anymore, the new one is https://developer.wordpress.org/reference/classes/wp_query/#category-parameters

titouanmathis commented on template/web/wp-content/themes/studiometa/app/Managers/ACFManager.php:
```diff
@@ -0,0 +1,84 @@
+namespace Studiometa\Managers;
+
+/** Class **/
+class ACFManager {
```
What about having a `ManagerInterface` defining the `run` method?

Ping @ptt-homme @perruche @lusimeon

perruche replied:
I guess it's probably a good idea, just a basic implementation like this is working, is it enough ? My experience with interfaces is pretty much non-existant

```php
/**
 * Interface for Managers
 */
namespace Studiometa\Interfaces;

/** Interface **/
interface ManagerInterface {
	public function run();
}
```

titouanmathis replied:
I think the namespace should be `Studiometa\Managers` as it is the interface for the manager classes. And we can add a documentation block before the `run` method, it will be readable by IDE and tools such as PHPStan.

**app/Managers/ManagerInterface.php**
```php
namespace Studiometa\Managers;

interface ManagerInterface {
  /**
   * ...
   * @return {void}
   */
  public function run();
}
```

**app/Managers/ThemeManager.php**
```php
namespace Studiometa\Managers;

class ThemeManager implements ManagerInterface {
  /**
   * @inheritdoc
   */
  public function run() {}
}
```

titouanmathis reviewed (changes requested):
Looks promising 👍

perruche committed:
Fix typing errors in ACFManager

perruche committed:
Add PHP warnings in PostTypeRepository methods if limit parameter exceed 100

perruche committed:
Use PostTypeRepository for archive and index

perruche committed:
Add an interface for Managers

titouanmathis commented on saofile.js:
```diff
@@ -88,7 +82,7 @@ module.exports = {
         type: 'move',
         patterns: {
-          'web/wp-content/themes/<%= slug %>': `web/wp-content/themes/${this.answers.slug}`,
+          'web/wp-content/themes/studiometa': `web/wp-content/themes/${this.answers.slug}`,
```
We can remove this line as we do not need to rename the `<%= slug =>` theme folder.

titouanmathis commented on template/composer.json:
```diff
@@ -91,6 +91,11 @@
   },
+  "autoload": {
+    "psr-4": {
+      "Studiometa\\": "web/wp-content/themes/studiometa/app"
```
👍

titouanmathis commented on template/web/wp-content/themes/studiometa/functions.php:
```diff
@@ -0,0 +1,68 @@
+			new \Studiometa\Managers\AssetsManager(),
+			new \Studiometa\Managers\CustomPostTypesManager(),
+			new \Studiometa\Managers\TaxonomiesManager(),
+		);
```
I think we can include the `ACFManager` directly in the `$managers` definition:

```php
$managers = array(
	new \Studiometa\Managers\WordPressManager(),
	new \Studiometa\Managers\TwigManager(),
	new \Studiometa\Managers\AssetsManager(),
	new \Studiometa\Managers\CustomPostTypesManager(),
	new \Studiometa\Managers\TaxonomiesManager(),
	<%_ if (acf) { _%>
	new \Studiometa\Managers\ACFManager(),
	<%_ } _%>
);
```

titouanmathis reviewed (changes requested):
Some minor tweaks left, and a fix in the `saofile.js` file, line 103, the`this.anwsers.slug` property does not exists anymore.

We can also add an action to remove the ACFManager file when ACF is not selected:

```js
if (!this.answers.features.includes('acf')) {
  actions.push({
    type: 'remove',
    files: 'web/wp-content/themes/studiometa/acf-json',
  });

  actions.push({
    type: 'remove',
    files: 'web/wp-content/themes/studiometa/app/Managers/ACFManager.php',
  });
}
```

perruche commented:
> We can also add an action to remove the ACFManager file when ACF is not selected:

I will also delete the `acf-json` folder, as the field groups will be written in php

perruche committed:
Update saofile

ptt-homme commented on template/_gitignore:
```diff
@@ -26,7 +26,7 @@ node_modules/
 
 # Add exception for the project's theme and plugins
-!/web/wp-content/themes/<%= slug %>
+!/web/wp-content/themes/studiometa
```
@studiometa/wordpress We have decided to use `studiometa` as the name of the theme. I'm wondering, why not use a default value `studiometa` and keep the ability to change it if needed!

ptt-homme commented on template/bin/source/wp-config-sample.php:
```diff
@@ -125,7 +125,7 @@
 
 /* Set default theme */
-define( 'WP_DEFAULT_THEME', '<%= slug %>');
+define( 'WP_DEFAULT_THEME', 'studiometa');
```
Code style 😛  `'studiometa' )`

ptt-homme commented on template/phpstan.neon:
```diff
@@ -45,4 +45,4 @@ parameters:
         # Please consider commenting ignores: issue URL or reason for ignoring
         - message: '#^Error(.+)thrown while looking(\s+)?for class Site\.#'
-          path: web/wp-content/themes/<%= slug %>/functions.php
+          path: web/wp-content/themes/studiometa/functions.php
```
@perruche Cette erreur existe encore ?

perruche committed:
Add a ManagerFactory to init all Managers

perruche committed:
Correct PHPCS errors

perruche committed:
Correct PHPStan errors

perruche committed:
Update the slug suggestion in saofile

perruche committed:
Fix a php error

perruche committed:
Add all dist files required by Studiometa\WP\Assets

- To prevent php error on project launch

perruche committed:
Fix a typeint

perruche committed:
Update PHPStan config file to scan WordPress plugins directory

- In order to avoid "function (plugin_function) not found errors"
- @see https://phpstan.org/user-guide/discovering-symbols

ptt-homme reviewed (approved):
Good for me!

ptt-homme reviewed (approved):
Good for me!

perruche committed:
Fix coding style in wp-config-sample
